feat: add a 'Scroll up for tutorial' button after scrolling past the instructions

trevinhofmann opened the pull request:
This pull request implements #172 by displaying a "Scroll up for tutorial" button when the user has scrolled far enough for the Tutorial component to no longer be visible. The visibility of the Tutorial component is detected using a new hook called `useIsVisible` which uses the [IntersectionObserver API](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API). The button has a slide and fade animation, and clicking it scrolls back to the top of the page to display the tutorial instructions.

Demo:

https://github.com/user-attachments/assets/4729455e-ca6a-4bb8-bd21-9f2682b3998f

raineorshine commented on src/components/Tutorial/TutorialScrollUpButton.tsx:
```diff
@@ -0,0 +1,39 @@
+   * Scrolls to the top of the window.
+   */
+  const scrollUp = useCallback(() => {
+    window.scrollTo({
```
Can you use https://github.com/cybersemics/em/blob/a33035e2076d82d8114459d662b0508ca353d686/src/device/scrollTo.ts? It's the same thing, but we abstract direct access to `window` to decouple the application logic from the browser API. This makes it easier to patch browser incompatibilities and mock functionality in tests.

trevinhofmann replied:
Definitely! I didn't see this at first.

Refactored here: 6019ce443289fb1f59919af916d833c0abf8d967

raineorshine commented on src/components/Tutorial/TutorialScrollUpButton.tsx:
```diff
@@ -0,0 +1,39 @@
+import { FC, useCallback } from 'react'
+import TutorialNavigationButton from './TutorialNavigationButton'
+
+interface TutorialScrollUpButtonProps {
+  show: boolean
+}
```
Let's inline this since it's so small.

trevinhofmann replied:
Makes sense. Refactored here: 6c6563bfd3a3c4bd77f942f963a4a14bdccfcdf7

raineorshine reviewed (changes requested):
Thanks! This looks great. The animation is exactly what I had in mind.

The one problem I found was that it does not work when the keyboard is up on mobile. This might be related to `position: fixed` on mobile Safari. I had the same problem with the toolbar. I have to switch it `position: absolute` and update `top` on scroll, but it's very choppy.

The other small issue is that there is some dead zone where the tutorial buttons are off screen, but the scroll up button is not shown. The scroll up button should be shown as soon as the Next/Prev buttons are no longer visible. I think we need a small offset to account for the padding in between the bottom edge of the Next/Prev buttons and the bottom edge of the `tutorial` div.

![Paste 2024-08-19 16 49 57](https://github.com/user-attachments/assets/5436f9e4-7ebf-4435-930e-7e4c374ab534)

trevinhofmann commented:
@raineorshine -- This is ready for review again, when you're available. Thanks!

>The one problem I found was that it does not work when the keyboard is up on mobile. This might be related to position: fixed on mobile Safari. I had the same problem with the toolbar. I had to switch it position: absolute and update top on scroll, but it's very choppy unfortunately.

Good find. After testing, I noticed that the iOS Safari viewport messes with both the `IntersectionObserver` and `position: 'fixed'` parts of this, _especially_ when the keyboard is open.

Quite unfortunate, but it seems the only solution compatible with iOS was to drop `IntersectionObserver` and manually calculate visibility from `getBoundingClientRect`: a345ed330b37b51a5448dc8ca67d0c30d41256f4

As you suggested, switching to `position: 'absolute'` and calculating the offset during scroll worked: 44885323ba41613a38cb2075cd0ec90234d8c282

I used CSS `calc` to combine the animation offset with the scroll offset at first, but realized that the animation caused additional lagginess during scrolling as the button would take longer to "catch up" to where the user scrolled. I fixed this by adding a separate div without animation which handles the scroll offset: e57df4658ab78a36fa743c67abb60f90d1b3730a

>The other small issue is that there is a dead zone where the tutorial buttons are off screen, but the scroll up button is not shown. Instead, the scroll up button should be shown as soon as the Next/Prev buttons are no longer visible. I think we need a small offset to account for the padding in between the bottom edge of the Next/Prev buttons and the bottom edge of the tutorial div.

We _could_ fix this with a hardcoded offset for the padding, but I'd rather avoid using a magic number which may vary from platform to platform, or over time as the app styling could change. Instead, I've switched the visibility `ref` from the entire tutorial panel to specifically the "next" tutorial navigation button: 9d6e996a3974574a2d7d9394f77df7a7e34b2393

Demo:

https://github.com/user-attachments/assets/fc35a782-53d4-4ee9-8d72-33d2c72a1ea9

raineorshine commented on src/hooks/useScrollPosition.ts:
```diff
@@ -0,0 +1,21 @@
+import { useEffect, useState } from 'react'
+
+/** Returns the vertical scroll position. */
+const useScrollPosition = (): number => {
```
We currently have a ministore that tracks scrollTop and throttles updates to 60fps: https://github.com/cybersemics/em/blob/main/src/stores/scrollTop.ts

I'm not sure if the throttling will add lag for us. But it certainly would be cleaner and more efficient to use the scrollTop ministore instead of a separate hook if possible. Can you try the ministore and see if there is a performance issue?

```js
const scrollTop = scrollTopStore.useState()
```

trevinhofmann replied:
Ah, my bad. I didn't notice the `scrollTopStore`.

Refactored: 5c38140d2a2e3dbe314612bfb23fd87fe8474dcb

raineorshine commented on src/hooks/useIsVisible.ts:
```diff
@@ -0,0 +1,37 @@
+
+  useEffect(() => {
+    checkVisibility()
+    window.addEventListener('scroll', checkVisibility)
```
This can definitely use the throttled scrollTop from scrollTopStore without any performance impact (see other comment)

trevinhofmann replied:
Yep! Switched here also: 5c38140d2a2e3dbe314612bfb23fd87fe8474dcb

raineorshine commented on src/components/Tutorial/TutorialScrollUpButton.tsx:
```diff
@@ -0,0 +1,45 @@
+
+/** A button to scroll up to the tutorial. */
+const TutorialScrollUpButton: FC<{ show: boolean }> = ({ show }) => {
+  const scrollPosition = useScrollPosition()
```
If we stick with subscribing directly to the scroll event, we should only enable it when the button is visible for performance reasons.

If we switch to the scrollTopStore, we can leave it as is.

trevinhofmann replied:
I've switched to use the `scrollTopStore` 👍

raineorshine commented on src/hooks/useIsVisible.ts:
```diff
@@ -0,0 +1,37 @@
+import { useEffect, useRef, useState } from 'react'
+
+/** Determines if an element is partially visible in the viewport. */
```
Should explain the purpose of the returned ref in the JSDOC.

trevinhofmann replied:
JSDoc Added: 37427088f917675d0fc014fa711d5212e4710c2a

raineorshine replied:
It appears `@param` comments do not work with VS Code's IntelliSense, but `/** ... */` comments do.

<img width="435" src="https://github.com/user-attachments/assets/20c87102-cc0d-4c80-9e9b-7d2b021d868c">

See: 9eef751896

raineorshine commented on src/components/Tutorial/TutorialNavigation.tsx:
```diff
@@ -20,7 +21,7 @@ import TutorialNavigationNext from './TutorialNavigationNext'
 
 // eslint-disable-next-line jsdoc/require-jsdoc
-const TutorialNavigation = ({ tutorialStep }: { tutorialStep: number }) => {
+const TutorialNavigation = ({ tutorialStep, nextRef }: { tutorialStep: number; nextRef: Ref<HTMLAnchorElement> }) => {
```
Let's add a JSDOC comment for `nextRef` explaining its purpose.

trevinhofmann replied:
Added the comment here: e2fbc16f6408d1397a80f37d44a58c546a933add

trevinhofmann commented:
Thank you @raineorshine -- Ready again for review 🎉

raineorshine commented on src/components/Tutorial/TutorialScrollUpButton.tsx:
```diff
@@ -0,0 +1,45 @@
+
+/** A button to scroll up to the tutorial. */
+const TutorialScrollUpButton: FC<{ show: boolean }> = ({ show }) => {
+  const scrollTop = scrollTopStore.useState()
```
For performance reasons, let's ensure the component is only re-rendered on scroll when visible.

You might try something like this:

```js
const scrollTop = scrollTopStore.useSelector(state => show ? 0 : state)
```

However my brief testing shows that the component is still getting re-rendered on scroll, so it may be coming from a parent also. Maybe the component needs `React.memo`?

trevinhofmann replied:
Rendering when scrolling and not-visible is necessary for the animation of the component. If we don't position it correctly before/after it being shown, then it does not transition to/from the correct location.

Changing the value of `scrollTop` to be `null` with that selector does not prevent the `scrollTopStore` from still triggering the re-render, and `React.memo` can only help with `props`, not the renders triggered within the component by the store.

Fortunately, I don't believe this should cause any worthwhile performance issues; if there were issues when the element is not displayed, then there would also be issues when the element is displayed. With the `scrollTopStore` already handling throttling and this element only being rendered during the brief tutorial, pursuing this particular case may be premature optimization.

raineorshine replied:
That's a fair point. I'm fine leaving it as-is then.

raineorshine commented on src/hooks/useIsVisible.ts:
```diff
@@ -0,0 +1,42 @@
+ * @returns An array containing the visibility state and the ref object of the element to check.
+ */
+const useIsVisible = <T extends HTMLElement>(initialValue = false): [boolean, React.RefObject<T>] => {
+  const scrollTop = scrollTopStore.useState()
```
You can also use the reactMinistore's `useEffect` directly:

```js
scrollTopStore.useEffect(checkVisibility)
```

trevinhofmann replied:
Very cool! Retrieving the `scrollTop` value only to use it as a hook dependency felt awkward, but I didn't realize `scrollTopStore.useEffect` existed. That's much nicer. Refactored: 84ce5aa6f6d67215bf3b1875c73a60aa29ab6612

raineorshine replied:
Yeah, there are a ton of React state management libraries out there, but honestly I've gotten so much mileage out of rolling my own [ministore](https://github.com/cybersemics/em/blob/2793615f9c268725f3ae9688ea1fa5250381a86a/src/stores/ministore.ts) and [react-ministore](https://github.com/cybersemics/em/blob/2793615f9c268725f3ae9688ea1fa5250381a86a/src/stores/react-ministore.ts) that are a mere 150 LOC combined.

Ministores can even be [composed](https://github.com/cybersemics/em/blob/2793615f9c268725f3ae9688ea1fa5250381a86a/src/components/LayoutTree.tsx#L89-L93).
